### ISSUE TITLE
The "cppdbg" provider strips the preLaunchTask from config

### ISSUE THREAD

**Issue by ZZH-Finalize** [microsoft/vscode-cpptools#9319]
### Environment

- OS and version: Windows_NT x64 6.1.7601
- VS Code: Code 1.67.1
- C/C++ extension: 1.10.2
- OS and version of remote machine (if applicable): none
- GDB / LLDB version: GNU gdb (GDB) (Cygwin 11.2-1) 11.2

### Bug Summary and Steps to Reproduce

Bug Summary:
I set a "preLaunchTask" to a shell task, but it just start debug without execute the task first.

Steps to reproduce:
1. Open a workspace by double click the `*.code-workspace` file
2. Set a task with `"type" : "shell"` in the `*.code-workspace` file
3. Add debug configuration in the `*.code-workspace` file and set `preLaunchTask` to the task which you creat in the last step
4. Start debug

### Debugger Configurations

```shell
{
    "folders": [
        {
            "path": "."
        }
    ],
    "tasks": {
        "version": "2.0.0",
        "tasks": [
            {
                "label": "build",
                "type": "shell",
                "command": "gcc",
                "args": [
                    "-o",
                    "main.exe",
                    "main.c"
                ],
                "group": {
                    "kind": "build",
                    "isDefault": true
                },
                "problemMatcher": "$gcc"
            },
            {
                "label": "test_task",
                "type": "shell",
                "command": "echo test_task run",
                "detail": "run this task should see a message",
                "group": {
                    "kind": "test",
                    "isDefault": true
                },
                "problemMatcher": []
            }
        ]
    },
    "launch": {
        "configurations": [
            {
                "name": "test",
                "type": "cppdbg",
                "request": "launch",
                "cwd": "${workspaceFolder}",
                "debugServerPath": "gdb",
                "program": "${workspaceFolder}/main.exe",
                "preLaunchTask": "test_task"
            }
        ],
        "compounds": []
    },
    "settings": {}
}
```

### Debugger Logs

```shell
gdb: GNU gdb (GDB) (Cygwin 11.2-1) 11.2
gdb: Copyright (C) 2022 Free Software Foundation, Inc.
gdb: License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
gdb: This is free software: you are free to change and redistribute it.
gdb: There is NO WARRANTY, to the extent permitted by law.
gdb: Type "show copying" and "show warranty" for details.
gdb: This GDB was configured as "x86_64-pc-cygwin".
gdb: Type "show configuration" for configuration details.
gdb: For bug reporting instructions, please see:
gdb: <https://www.gnu.org/software/gdb/bugs/>.
gdb: Find the GDB manual and other documentation resources online at:
gdb:     <http://www.gnu.org/software/gdb/documentation/>.
gdb: 
gdb: For help, type "help".
gdb: Type "apropos word" to search for commands related to "word".
=thread-group-added,id="i1"
GNU gdb (GDB) (Cygwin 11.2-1) 11.2
Copyright (C) 2022 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Type "show copying" and "show warranty" for details.
This GDB was configured as "x86_64-pc-cygwin".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<https://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
    <http://www.gnu.org/software/gdb/documentation/>.

For help, type "help".
Type "apropos word" to search for commands related to "word".
Warning: Debuggee TargetArchitecture not detected, assuming x86_64.
=cmd-param-changed,param="pagination",value="off"
=library-unloaded,id="/cygdrive/c/Windows/system32/apphelp.dll",target-name="/cygdrive/c/Windows/system32/apphelp.dll",host-name="/cygdrive/c/Windows/system32/apphelp.dll",thread-group="i1"
[New Thread 5388.0x1208]
[New Thread 5388.0x1314]
[New Thread 5388.0x7e4]

Thread 1 "main" hit Breakpoint 1, 0x000000010040108d in main ()
Loaded '/cygdrive/c/Windows/SYSTEM32/ntdll.dll'. Symbols loaded.
Loaded '/cygdrive/c/Windows/system32/kernel32.dll'. Symbols loaded.
Loaded '/cygdrive/c/Windows/system32/KERNELBASE.dll'. Symbols loaded.
Loaded '/usr/bin/cygwin1.dll'. Symbols loaded.
Loaded '/cygdrive/c/Windows/system32/advapi32.dll'. Symbols loaded.
Loaded '/cygdrive/c/Windows/system32/msvcrt.dll'. Symbols loaded.
Loaded '/cygdrive/c/Windows/SYSTEM32/sechost.dll'. Symbols loaded.
Loaded '/cygdrive/c/Windows/system32/RPCRT4.dll'. Symbols loaded.
Loaded '/cygdrive/c/Windows/system32/cryptbase.dll'. Symbols loaded.
Loaded '/cygdrive/c/Windows/system32/netapi32.dll'. Symbols loaded.
Loaded '/cygdrive/c/Windows/system32/netutils.dll'. Symbols loaded.
Loaded '/cygdrive/c/Windows/system32/srvcli.dll'. Symbols loaded.
Loaded '/cygdrive/c/Windows/system32/wkscli.dll'. Symbols loaded.
Loaded '/cygdrive/c/Windows/system32/LOGONCLI.DLL'. Symbols loaded.
Loaded '/cygdrive/c/Windows/system32/SAMCLI.DLL'. Symbols loaded.
Loaded '/cygdrive/c/Windows/system32/SAMLIB.dll'. Symbols loaded.
Loaded '/cygdrive/c/Windows/system32/ws2_32.dll'. Symbols loaded.
Loaded '/cygdrive/c/Windows/system32/NSI.dll'. Symbols loaded.
Loaded '/cygdrive/c/Windows/system32/mswsock.dll'. Symbols loaded.
Loaded '/cygdrive/c/Windows/system32/user32.dll'. Symbols loaded.
Loaded '/cygdrive/c/Windows/system32/GDI32.dll'. Symbols loaded.
Loaded '/cygdrive/c/Windows/system32/LPK.dll'. Symbols loaded.
Loaded '/cygdrive/c/Windows/system32/USP10.dll'. Symbols loaded.
Loaded '/cygdrive/c/Windows/system32/IMM32.DLL'. Symbols loaded.
Loaded '/cygdrive/c/Windows/system32/MSCTF.dll'. Symbols loaded.
test
[Thread 5388.0x2c0 exited with code 0]
[Thread 5388.0x1208 exited with code 0]
[Thread 5388.0x1314 exited with code 0]
[Inferior 1 (process 5388) exited normally]
The program 'F:\vscode-test/main.exe' has exited with code 0 (0x00000000).
```

### Additional Information

This is the reply that i get it from vscode deleloper

> in the `DebugService` `createSession` call, the `config` is passed in with the correct `preLaunchTask`.
> 
> ![Screen Shot 2022-05-16 at 9 54 50 PM](https://user-images.githubusercontent.com/29464607/168732651-2c60a6cd-d079-45fa-a060-18cb8eff9c93.png)
> 
> `DebugConfigurationManager` resolves the `config` via `resolveConfigurationByProviders`
> 
> the `cppdbg` provider here strips the `preLaunchTask` from `config` such that it's missing from the `result` object that gets returned ![Screen Shot 2022-05-16 at 9 58 26 PM](https://user-images.githubusercontent.com/29464607/168732792-fbbb7861-cc58-4451-9f3a-8e3e38f14799.png) ![Screen Shot 2022-05-16 at 9 55 30 PM](https://user-images.githubusercontent.com/29464607/168732795-3a757200-14a7-41f9-a29a-238c46d379bc.png)

https://user-images.githubusercontent.com/33342565/168737207-757512e9-9dd8-4066-a084-6828d1d8e3b1.mov

**Comment by ZZH-Finzlize** [comment]
@sean-mcmanus could you look into the problem? i'm very sure this is a bug caused by `ms-vscode.cpptools`

**Comment by ZZH-Finzlize** [comment]
@sean-mcmanus  I guess i find the bug. These code here, it just judge the envirment is a folder or not, but actually vscode suport workspace, it's neither a folder nor a single file. So i guess when i use workspace, these code will treat my workspace as a single file, then, it will delete the "preLaunchTask" and cause this bug.

**Comment by elahehrashedi** [comment]
The main cause of this issue is that you are defining the tasks in your  `*.code-workspace` file and currently, our extension is not looking for the tasks outside of the tasks.json file. 

As a workaround, you can move all tasks to the `tasks.json`. 

You don't have to move the launch configuration, the extension can reads that info from  `*.code-workspace` file.

I can see that you have changed the type of the task from `cppbuild` to `shell`. the `*.code-workspace` only supports shell tasks, correct? Is that the reason you modified the type of the task? You can modify that back to `cppbuild`. Let me know if that approach works for you.

**Comment by elahehrashedi** [comment]
The extension can not define tasks in code workspace files yet (https://github.com/microsoft/vscode/issues/93149), which can be a good reason not to scan this file to extract the tasks. This issue can be a feature request to scan for "shell" tasks everywhere in the workspace.

**Comment by gerardonetzyn** [comment]
I'm experiencing the same problem: VSCode 1.66 used to launch the prelaunchTask defined in same workspace file. 1.67  doesn't. I use this feature to build my CPP projects before debugging them in Linux.

**Comment by elahehrashedi** [comment]
@gerardonetzyn which version of the extension are you using?

**Comment by gerardonetzyn** [comment]
VS Code: Code 1.67.2
C/C++ extension: 1.10.2 (Pre-Release)

**Comment by elahehrashedi** [comment]
If this seems like a feature that users have been using, we can consider it as a feature request.
You can still see the shell tasks and other tasks when running `Tasks: Run Task`. and you can run your build tasks before debugging.
The launch info is only needed for debugging. 
We were not searching for tasks in `*.code-workspace` for debugging in previous versions either.

**Comment by elahehrashedi** [comment]
@gerardonetzyn so your issue has been resolved?

**Comment by gerardonetzyn** [comment]
@elahehrashedi  Kind of? If using the extensions' pre-release version, the prelaunch task in workspace is NOT launched. If using the release version, prelaunch taks is launched. So, my take is that once the pre-release code makes it to the release code, I won't be able to prelanch tasks anymore. Does that make sense?

**Comment by elahehrashedi** [comment]
@gerardonetzyn 

can you please share your tasks and launch configurations? 

Are you using `F5` or `run and debug` play button to launch the debug session?

**Comment by ZZH-Finzlize** [comment]
> The extension can not define tasks in code workspace files yet ([microsoft/vscode#93149](https://github.com/microsoft/vscode/issues/93149)), which can be a good reason not to scan this file to extract the tasks. This issue can be a feature request to scan for "shell" tasks everywhere in the workspace.

This is not the same problem, if task type is `shell`, it will work in "*.vscode-workspace" file.
![image](https://user-images.githubusercontent.com/33342565/169627248-7eacc678-1b93-4e88-b602-64162b20805c.png)

 As a extention , you don't need to scan any file, the configure is aleady given by vscode, all you need to do is do not delete `prelaunchTask` from the config. This is changed at Pre-Release version, but it works well before, so i didn't think this is a feature, it's a function that already have.

> @meganrogge You're right, i rewrite those code in my local file, and the problem is solved. ![image](https://user-images.githubusercontent.com/33342565/169235352-129c5ba6-a3c8-4eb9-ab06-207dc9e96bd5.png)
> 
> This is my debug log ![image](https://user-images.githubusercontent.com/33342565/169236165-38aa3f11-ff09-4645-adf7-a5772eb9bc29.png)
> 
> Your answer is really helpful

**Comment by gerardonetzyn** [comment]
@elahehrashedi I start my debug sessions by pressing F5 or using the run-and-debug play button.

This is my launch configuration in my workspace file:

```
{
  "launch": {
    "version": "0.2.0",
      "configurations": [
      {
        "name": "Debug",
        "program": "${workspaceFolder}/out-debug/NzDemo/NzDemo",
        "cwd": "${workspaceFolder}",
        "preLaunchTask": "Build",
        "type": "cppdbg",
        "request": "launch",
        "stopAtEntry": false,
        "externalConsole": false,
        "MIMode": "gdb",
        "sourceFileMap": {
          "${workspaceFolder}/out-debug/NzDemo": "${workspaceFolder}/NzDemo"
        }
      },
    ],
  },
  
  "tasks": {
    "version": "2.0.0",
    "tasks": [
      {
        "label": "Build",
        "type": "shell",
        "problemMatcher": {
          "base": "$gcc", 
          "fileLocation": [ "relative", "${workspaceFolder}/1/2" ]
        },
        "command": "clear; echo '**** Building NzDemo *****';tup ${workspaceFolder}/out-debug/NzDemo --quiet",
        "group": {
          "kind": "build",
          "isDefault": true
        }
      },      
    ],
  },
}

```

**Comment by sean-mcmanus** [comment]
This was fixed a long time ago (1.10).
